File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/gift/view/animation/manager/GiftAnimationManager.java
package com.trtc.uikit.livekit.component.gift.view.animation.manager;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

public class GiftAnimationManager {

    private static final String TAG = "GiftAnimationManager";

    private static final int MAX_CACHE_SIZE               = 3;
    private static final int ANIMATION_PREPARE_TIMEOUT_MS = 2000;

    private final Handler                  mMainHandler     = new Handler(Looper.getMainLooper());
    private final List<GiftAnimationModel> mGiftPrepareList = new ArrayList<>();

    private AnimationPlayer mAnimationPlayer;
    private boolean         mIsPlaying;

    private final Runnable mPrepareTimeoutTask = () -> {
        Log.w(TAG, "Prepare timeout");
        onFinished(-1);
    };

    public GiftAnimationManager() {
    }

    public void setPlayer(AnimationPlayer player) {
        mAnimationPlayer = player;
        mAnimationPlayer.setCallback(this::onFinished);
    }

    public void add(GiftAnimationModel gift) {
        if (gift.isFromSelf) {
            if (mGiftPrepareList.isEmpty()) {
                mGiftPrepareList.add(gift);
            } else {
                // Add after the last gift which is from self.
                int lastIndex = 0;
                for (int i = mGiftPrepareList.size() - 1; i > 0; i--) {
                    GiftAnimationModel giftModel = mGiftPrepareList.get(i);
                    if (giftModel.isFromSelf) {
                        lastIndex = i;
                        break;
                    }
                }
                mGiftPrepareList.add(lastIndex + 1, gift);
            }
        } else {
            mGiftPrepareList.add(gift);
        }
        // remove redundant gift
        if (mGiftPrepareList.size() == MAX_CACHE_SIZE + 1) {
            int removeIndex = 1;
            for (int i = 1; i < mGiftPrepareList.size(); i++) {
                GiftAnimationModel giftModel = mGiftPrepareList.get(i);
                if (!giftModel.isFromSelf) {
                    removeIndex = i;
                    break;
                }
            }
            mGiftPrepareList.remove(removeIndex);
        }
        if (mGiftPrepareList.size() == 1 && !mIsPlaying) {
            preparePlay(mGiftPrepareList.remove(0));
        }
    }

    public void startPlay(GiftAnimationModel model) {
        Log.i(TAG, "startPlay:" + model.gift.resourceUrl);
        mMainHandler.removeCallbacks(mPrepareTimeoutTask);
        if (mAnimationPlayer != null) {
            mAnimationPlayer.startPlay(model);
        }
    }

    public void stopPlay() {
        Log.i(TAG, "stopPlay");
        if (mAnimationPlayer != null) {
            mAnimationPlayer.stopPlay();
        }
    }

    private void preparePlay(GiftAnimationModel giftModel) {
        Log.i(TAG, "preparePlay:" + giftModel.gift.resourceUrl);
        mIsPlaying = true;
        if (mAnimationPlayer != null) {
            mAnimationPlayer.preparePlay(giftModel);
        }
        mMainHandler.postDelayed(mPrepareTimeoutTask, ANIMATION_PREPARE_TIMEOUT_MS);
    }

    private void onFinished(int error) {
        Log.i(TAG, "onFinished:" + error);
        mIsPlaying = false;
        if (!mGiftPrepareList.isEmpty()) {
            preparePlay(mGiftPrepareList.remove(0));
        }
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/gift/view/animation/manager/GiftAnimationManager.kt
package com.trtc.uikit.livekit.component.gift.view.animation.manager

import android.util.Log
import com.trtc.uikit.livekit.common.LiveKitLogger
import com.trtc.uikit.livekit.common.LiveKitLogger.Companion.getComponentLogger
import com.trtc.uikit.livekit.component.gift.viewmodel.GiftModel

class GiftAnimationManager {
    private val logger: LiveKitLogger = getComponentLogger("GiftAnimationManager")
    private val giftPrepareList: MutableList<GiftModel> = ArrayList()
    private var animationPlayer: AnimationPlayer? = null
    private var isPlaying = false

    fun setPlayer(player: AnimationPlayer?) {
        animationPlayer = player
        animationPlayer?.setCallback { error: Int -> this.onFinished(error) }
    }

    fun add(gift: GiftModel) {
        if (gift.isFromSelf) {
            if (giftPrepareList.isEmpty()) {
                giftPrepareList.add(gift)
            } else {
                var lastIndex = 0
                for (i in giftPrepareList.size - 1 downTo 1) {
                    val giftModel = giftPrepareList.get(i)
                    if (giftModel.isFromSelf) {
                        lastIndex = i
                        break
                    }
                }
                giftPrepareList.add(lastIndex + 1, gift)
            }
        } else {
            giftPrepareList.add(gift)
        }
        if (giftPrepareList.size == MAX_CACHE_SIZE + 1) {
            var removeIndex = 1
            for (i in giftPrepareList.indices) {
                val giftModel = giftPrepareList[i]
                if (!giftModel.isFromSelf) {
                    removeIndex = i
                    break
                }
            }
            giftPrepareList.removeAt(removeIndex)
        }
        if (giftPrepareList.size == 1 && !isPlaying) {
            preparePlay(giftPrepareList.removeAt(0))
        }
    }

    fun startPlay(model: GiftModel) {
        logger.info( "startPlay:${model.gift?.resourceURL},isPlaying:$isPlaying")
        animationPlayer?.startPlay(model)
    }

    fun stopPlay() {
        logger.info( "startPlay:isPlaying:$isPlaying")
        isPlaying = false
        animationPlayer?.stopPlay()
    }

    private fun preparePlay(model: GiftModel) {
        logger.info( "preparePlay:${model.gift?.resourceURL},isPlaying:$isPlaying")
        isPlaying = true
        animationPlayer?.preparePlay(model)
    }

    private fun onFinished(error: Int) {
        Log.i(TAG, "onFinished:$error")
        isPlaying = false
        if (!giftPrepareList.isEmpty()) {
            preparePlay(giftPrepareList.removeAt(0))
        }
    }

    companion object {
        private const val TAG = "GiftAnimationManager"
        private const val MAX_CACHE_SIZE = 3
    }
}